### PR TITLE
REST API - Entities API docs

### PR DESCRIPTION
**Pull request by mariodavid**
This PR contains the Entities API section of the REST API

### CODE REVIEW

File: content/modules/rest/pages/quick-start.adoc [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+Let’s consider that we want to retrieve information about orders in the application through the REST API.
+
+First, add generic REST to your project according to the xref:index.adoc#installation[installation] section.
+
+== Create REST enabled Role
+
+
+The first thing is to configure security according to <<security,Security>> section:
+
+* Create a role that includes `rest.enabled` specific permission and grants permissions to read the user entity
```
Comment (knstvk) [inline content/modules/rest/pages/quick-start.adoc]: Dot, even if it's a bullet list item.

File: content/modules/rest/pages/quick-start.adoc [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+* Assign the created role to the user that you want to use (like `admin`)
+
+Start the application and test the REST API using the `curl` command-line tool.
+
+== Authenticate as a User
+
+As the first request, you need to perform a login or `Authentication request`. This happens through the following command:
```
Comment (knstvk) [inline content/modules/rest/pages/quick-start.adoc]: https://www.google.com/search?q="This%20happens%20through%20the%20following%20command"
Zero results.
"Use the following command"?

File: content/modules/rest/pages/quick-start.adoc [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+{
+  "access_token": "CXE0w/9cOsnpSo8v2jEDoI8Qa3Y=",
+  "token_type": "bearer",
+  "refresh_token": "Hh2xCuZ7fgd35obagEBNGevF4ws=",
+  "expires_in": 31535999,
+  "scope": "rest-api",
+  "OAuth2.SESSION_ID": "5C46CDF266E8C8C15372887830B74F59"
+}
+----
+The attribute `access_token` contains the token that you can use for further request as part of the `Authorization` Header.
```
Comment (knstvk) [inline content/modules/rest/pages/quick-start.adoc]: "further requests"

File: content/modules/rest/pages/quick-start.adoc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+
+== Retrieve Entity List
+Now that you performed a successful login on the API, the generic REST API endpoints can be used to perform
```
Comment (knstvk) [inline content/modules/rest/pages/quick-start.adoc]: "can be used to request the list of users"

File: content/modules/rest/pages/generic-functionality.adoc [hunk 5]
```diff
@@ -0,0 +1,39 @@
+= Generic Functionality
+
+The Generic REST API comes with different built-in functionalities to interact with a Jmix application. The following generic functionalities are available as part of the built-in offering:
```
Comment (knstvk) [inline content/modules/rest/pages/generic-functionality.adoc]: Plz remove the second sentence ("The following generic ..."), it repeats the first one

File: content/modules/rest/pages/entities-api.adoc [hunk 6]
```diff
@@ -0,0 +1,724 @@
+= Entities API
+
+The Entities API provides different capabilities to interact with entity instances. This API mainly consists of CRUD operations that can be performed on the entities. The ability to load entity data is extended by different ways of filtering and JPQL.
+
+NOTE: When performing operations on entities through this generic API, the regular Jmix security permission are applied from the user obtaining the authentication token.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "security permissions"
And maybe "from the user **obtained** the authentication token"?

File: content/modules/rest/pages/entities-api.adoc [hunk 7]
```diff
@@ -0,0 +1,724 @@
...
+
+
+== Load Entities
+
+The Entities API allows different ways to load entities through the API:
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Plz avoid using "allows" without "whom" - I heard about it from our English colleagues. 
"The Entities API allows you to load entities in different ways"

File: content/modules/rest/pages/entities-api.adoc [hunk 8]
```diff
@@ -0,0 +1,724 @@
...
+
+* <<Load Entity by ID,Load Entity by ID>> allows to load a single entity by its unique identifier (ID)
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "allows you to load" or "lets you load" or "enables loading".
And don't forget dots at the end.

File: content/modules/rest/pages/entities-api.adoc [hunk 9]
```diff
@@ -0,0 +1,724 @@
...
+* <<Load Entities via Search Filter,Load Entities via Search Filter>> allows specifying filters to only load entities that match those criteria
+
+* <<Load Entities via JPQL,Load Entities via JPQL>> allows pre-configuring JPQL queries in the application and use them in the API as named queries
+
+
+=== Load Entity by ID
+
+The first way of loading an entity via the Entities API is by loading it via its ID. The corresponding endpoint for this is the Load Entity endpoint `entities/:entityName/:entityId`.
+
+The `:entityName` Path parameter defines which entity type it is. The value is defined in the entity definition:
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "path"
Why capitalize?

File: content/modules/rest/pages/entities-api.adoc [hunk 10]
```diff
@@ -0,0 +1,724 @@
...
+@Entity(name = "rstex11_Order") //<1>
+public class Order {
+    // ...
+}
+----
+<1> the name attribute value `rstex11_Order` of the `@Entity` JPA annotation indicates the `entityName` parameter in the Entities API
+
+The Load Entity endpoint returns a single instance in case it is found by its identifier. Otherwise, HTTP status code `404 - Not found` is returned.
+
+In the example the Order with the ID `21021f78-edac-224b-e6f8-6e71e02a0f0d` is loaded. Furthermore, in this example the optional URL query parameter `fetchPlan` is used to define which attributes to return as part of the response.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "In the example below"

File: content/modules/rest/pages/entities-api.adoc [hunk 11]
```diff
@@ -0,0 +1,724 @@
...
+    "id": "0826806e-6074-90fa-f241-564b5c94d018",
+    "createdDate": "2021-02-06T12:00:53.72",
+    "createdBy": "admin",
+    "lastModifiedDate": "2021-02-06T12:00:53.72",
+    "name": "Sidney Chandler",
+    "version": 1
+  }
+}
+----
+<1> some metadata about the entity instance is returned as JSON keys (`entityName`, `_instanceName` and `id`)
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Start from capital letter, end with dot.

File: content/modules/rest/pages/entities-api.adoc [hunk 12]
```diff
@@ -0,0 +1,724 @@
...
+<2> the business attributes of the entity are returned as JSON keys
+<3> due to the fetch plan `order-with-details` used in the URL, additional reference attributes are also included like `lines` and `customer`
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Avoid passive voice where possible. There is too much of it here.
"The usage of `order-with-details` fetch plan ensures that additional reference attributes like `lines` and `customer` are also included."

File: content/modules/rest/pages/entities-api.adoc [hunk 13]
```diff
@@ -0,0 +1,724 @@
...
+
+=== Load List of Entities
+
+The next way of loading entities through the Entities API is to load a list of entities via the Load Entity List API: `/entities/:entityName`. Generally this API allows loading all entity instances of a given type. The API includes pagination, sorting and fetch plans.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Sounds too wordy to me, what about to simplify:
"You can load a list of entities of any type using the Load Entity List API endpoint: `/entities/:entityName`. This API includes pagination, sorting and fetch plans."

File: content/modules/rest/pages/entities-api.adoc [hunk 14]
```diff
@@ -0,0 +1,724 @@
...
+    "createdDate": "2021-02-06T12:02:05.328",
+    "createdBy": "admin",
+    "lastModifiedDate": "2021-02-06T12:02:05.328",
+    "name": "Shelby Robinson",
+    "version": 1
+  }
+]
+----
+
+It is also possible to further control the behavior of the API by using the following URL query parameter.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "parameters:"

File: content/modules/rest/pages/entities-api.adoc [hunk 15]
```diff
@@ -0,0 +1,724 @@
...
+
+[NOTE]
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Why NOTE?

File: content/modules/rest/pages/entities-api.adoc [hunk 16]
```diff
@@ -0,0 +1,724 @@
...
+.Load Entity List API Query Parameters
+====
+[horizontal]
+fetchPlan:: name of an entity types xref:data-access:fetch-plans.adoc[fetch plan], that will determine the JSON attributes each entity contains `(String)`
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "name of an entity's fetch plan"

File: content/modules/rest/pages/entities-api.adoc [hunk 17]
```diff
@@ -0,0 +1,724 @@
...
+limit:: the number of entities to be returned by the API `(int)`
+offset:: the position of the first returned entity `(int)`
+sort:: an entity attribute that will be used for sorting. `+attribute` or `attribute` for ascending order, `-attribute` for descending order `(String)`
+====
+
+In the following example the third "page" of 5 `Order` entities should be loaded (the 10 - 15 entity) sorted by the `date` attribute containing all attributes of the `order-with-details` fetch plan.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: How I would write it without passive voice:
"The following example demonstrates how to load the third "page" of 5 `Order` entities (the 10 - 15 entity) sorted by the `date` attribute and containing all attributes of the `order-with-details` fetch plan."

File: content/modules/rest/pages/entities-api.adoc [hunk 18]
```diff
@@ -0,0 +1,724 @@
...
+    }
+  }
+]
+----
+
+NOTE: Every entity in the response has a `_entityName` attribute with the name of the entity, and an `_instanceName` attribute with the xref:data-model:entities.adoc#instance-name[Instance Name] of the entity.
+
+=== Load Entities via Search Filter
+
+It is possible to specify filter criteria when fetching entities. For this use-case the Entity Search Endpoint is available via `/entities/:entityName/search`.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: I can't help but simplify your introductory sentences:
"You can specify filter criteria when loading entities using the Entity Search Endpoint: `/entities/:entityName/search`."

File: content/modules/rest/pages/entities-api.adoc [hunk 19]
```diff
@@ -0,0 +1,724 @@
...
+
+Both HTTP methods `GET` and `POST` are possible when interacting with the search endpoint. In both cases the filter criterion has to be provided as part of the request.
+
+The filter definition is a JSON structure that contains a set of conditions. A condition contains of the attributes' `property`, `operator` and `value`.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "A condition consists of the `property`, `operator` and `value` attributes."

File: content/modules/rest/pages/entities-api.adoc [hunk 20]
```diff
@@ -0,0 +1,724 @@
...
+
+[NOTE]
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Why NOTE?

File: content/modules/rest/pages/entities-api.adoc [hunk 21]
```diff
@@ -0,0 +1,724 @@
...
+[horizontal]
+property:: the entity attribute that is being filtered on (like `amount` on the Order entity).
++
+In case the attribute is a reference to another entity, it can also be a property path like `customer.name`
+
+operator:: the filter operator. The following values are possible (depending on the property type):
++
+`=`, `>`, `>=`, `<`, `&lt;=`, `<>`, `startsWith`, `endsWith`, `contains`, `doesNotContain`, `in`, `notIn`, `notEmpty`, `isNull`
+
+value:: the value to search for. Value is required expect for the operators `notEmpty` and `isNull`
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Probably "Value is not required for the `notEmpty` and `isNull` operators."

File: content/modules/rest/pages/entities-api.adoc [hunk 22]
```diff
@@ -0,0 +1,724 @@
...
+    {
+      "property": "booleanField",
+      "operator": "=",
+      "value": true
+    }
+  ]
+}
+----
+
+This is a representation of the Filter criterion: `((stringField = stringValue) OR (intField > 100) AND (booleanField = true))`.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: `((` and `))` are not rendered for some reason, probably need escaping

File: content/modules/rest/pages/entities-api.adoc [hunk 23]
```diff
@@ -0,0 +1,724 @@
...
+            /entities
+            /rstex11_Order
+            /search
+            ?filter={"conditions":[{"property":"customer.name","operator":"contains","value":"Shelby"}]}
+----
+
+[IMPORTANT]
+.URI Encoding
+====
+The HTTP URI standard only allows ASCII characters as part of the URI / URL. When using URL Query parameters for the filter definition, the JSON definition has to be URL encoded in order to match this requirement. This is also true for the `value` data that normally contains user-input.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "user input"

File: content/modules/rest/pages/entities-api.adoc [hunk 24]
```diff
@@ -0,0 +1,724 @@
...
+
+As there is also a practical limit of the URI length that can cause problems for big filter definitions, the `POST` method for filtering entities should be preferred as those restrictions do not apply.
+====
+
+=== Load Entities via JPQL
+
+Another alternative to load entities from the application is to use predefined JPQL queries. The Entity Query Endpoint `/queries/:entityName/:queryName` is responsible for providing this capability. Queries can contain a list of parameters, that needs to be provided by the client. Additionally, the endpoint contains the same general parameters for pagination, fetchPlans etc.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "fetch plans"

File: content/modules/rest/pages/entities-api.adoc [hunk 25]
```diff
@@ -0,0 +1,724 @@
...
+
+
+[TIP]
+.When to use JPQL vs. Search Filter?
+====
+Jmix provides various ways to load entity data in a generic way. Using pre-defined JPQL queries can be used when the search filter are not advanced enough to express the filter criterion. Also in case the parameter should be pre-defined and not be adjustable by the API client.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "... Use pre-defined JPQL queries when the search filter is not advanced enough ..."

File: content/modules/rest/pages/entities-api.adoc [hunk 26]
```diff
@@ -0,0 +1,724 @@
...
+====
+
+==== JPQL Query Configuration
+
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc:375]: Plz try to use less passive voice in this section. There is too many of "need to be ..." and the like.

File: content/modules/rest/pages/entities-api.adoc [hunk 27]
```diff
@@ -0,0 +1,724 @@
...
+            <param name="customerName" type="java.lang.String"/>
+        </params>
+    </query>
+</queries>
+----
+
+A query needs to have a unique `name` value as well as an `entity` reference. The combination of `name` and `entity` needs to be unique. Also, a `fetchPlan` needs to be referenced to indicate which entity attributes are returned.
+
+In the `<jpql>` tag the actual query is configured. Parameters are referenced via their name prefixed with a colon like `:customerName`.
+The parameters itself need to be listed within the `params` tag defining their name and Java type.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "parameters themselves"?

File: content/modules/rest/pages/entities-api.adoc [hunk 28]
```diff
@@ -0,0 +1,724 @@
...
+        /rstex11_Order
+        /ordersByDate
+        ?orderDate=2020-02-02
+----
+
+
+[IMPORTANT]
+.URI Encoding
+====
+Similar to the Entity Search Endpoint the URL should only contain ASCII characters. This means the values of the parameters need to be URL encoded, as those values normally represent direct user input and therefore it cannot be ensured that non-ascii characters are used.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Perhaps reference to Entity Search Endpoint is not relevant here, just "The URL should ..."

File: content/modules/rest/pages/entities-api.adoc [hunk 29]
```diff
@@ -0,0 +1,724 @@
...
+<queries xmlns="http://jmix.io/schema/rest/queries">
+    <query name="ordersByIds" entity="rstex11_Order" fetchPlan="order-with-details">
+        <jpql><![CDATA[select e from rstex11_Order e where e.id in :ids]]></jpql>
+        <params>
+            <param name="ids" type="java.util.UUID[]"/> // <1>
+        </params>
+    </query>
+</queries>
+----
+<1> the `ids` parameter is marked as collection of `UUID` type
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Start from capital letter and end with dot.
Plz fix it also for all similar cases.

File: content/modules/rest/pages/entities-api.adoc [hunk 30]
```diff
@@ -0,0 +1,724 @@
...
+  "ids": [
+    "41aae331-b46b-85ee-b0bc-2de8cbf1ab86",
+    "21021f78-edac-224b-e6f8-6e71e02a0f0d"
+  ]
+}
+----
+
+== Create Entities
+
+The Entities API allows creating entities through the Create Entity endpoint. An entity can be created via performing a `POST` request against the endpoint: `/entities/:entityName`.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: My simplified version:
"The Entities API lets you create entities by performing a `POST` request against the `/entities/:entityName` endpoint."

File: content/modules/rest/pages/entities-api.adoc [hunk 31]
```diff
@@ -0,0 +1,724 @@
...
+
+The request body contains a JSON object with the attributes of the entity.
+
+NOTE: Attributes from xref:data-model:entities.adoc#traits[entity traits] like `id` or `createdBy` should not be part of the request. Those attributes are automatically added by Jmix while storing the entity.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc:481]: Actually, if the entity's id attribute has no `JmixGeneratedValue` or `GeneratedValue` annotations, the client has to provide an id value.

File: content/modules/rest/pages/entities-api.adoc [hunk 32]
```diff
@@ -0,0 +1,724 @@
...
+}
+----
+
+Alternatively it is possible to define which attributes should be returned once the entity is created. This can be achieved through the URL query parameter `responseFetchPlan`. E.g. `/entities/rstex11_Order?responseFetchPlan=order-with-details` would return the complete order with all its details containing order lines, customer references etc.
+
+TIP: The HTTP response Header `Location` indicates also the URL to the newly created entity instance for further operations (like fetching, updating or deleting).
+
+=== Entity Validation
+
+When an entity is created or updated by default the regular entity validation is active and enforced. This means that invalid input (according to the entity validation annotations) from the API is rejected with the HTTP response status code `400 - Bad Request`.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "When an entity is created or updated, the regular entity validation is active and enforced by default."

File: content/modules/rest/pages/entities-api.adoc [hunk 33]
```diff
@@ -0,0 +1,724 @@
...
+
+The API returns detailed error messages for each validation violation in the following form of an JSON array. Each entry follows the structre
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "a JSON"
"the structure:"

File: content/modules/rest/pages/entities-api.adoc [hunk 34]
```diff
@@ -0,0 +1,724 @@
...
+
+[NOTE]
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Why NOTE?

File: content/modules/rest/pages/entities-api.adoc [hunk 35]
```diff
@@ -0,0 +1,724 @@
...
+    "path": "customer",
+    "invalidValue": null
+  }
+]
+----
+
+
+=== Association Attributes
+
+When the entity to create should be linked to another, already existing entity it is required to reference the other entity in the request. In the Order example, a `customer` needs to be referenced when an Order is created. This is done through a JSON object containing only the ID of the existing entity. Jmix performs a lookup of the customer by the provided ID from the JSON request and links the customer to the new order.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: "... another, already existing entity it, is required to ..."

File: content/modules/rest/pages/entities-api.adoc [hunk 36]
```diff
@@ -0,0 +1,724 @@
...
+
+    @Composition
+    @OneToMany(mappedBy = "order")
+    private List<OrderLine> lines;
+
+    // ...
+}
+----
+
+When an order is created through the API it is possible to directly create its order lines as part of the request. In this case all the attributes of the child entity need to be provided, in order to create it. The relationship from the parent to the child does not need to be additionally referenced. Putting the child entity into the JSON array is enough to establish the connection.
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Needs some punctuation.
"... created through the API, it is possible ... "
"In this case, all ..."

**Review by knstvk — CHANGES_REQUESTED** [review]
Excellent work! You are making a comprehensive and well-structured guide. 
My comments are mostly about wording and typos.

File: content/modules/rest/pages/entities-api.adoc (hunk 15 above)
Comment (mariodavid) [reply content/modules/rest/pages/entities-api.adoc]: i just found that visually appealing to somehow render it differently. I think the Admonitions from Asciidoc are a good way of doing that. Are you questioning the particular Admonitions type (Note) or the Admonition in general?

File: content/modules/rest/pages/entities-api.adoc (hunk 20 above)
Comment (mariodavid) [reply content/modules/rest/pages/entities-api.adoc]: see explanation above

File: content/modules/rest/pages/entities-api.adoc (hunk 31 above)
Comment (mariodavid) [reply content/modules/rest/pages/entities-api.adoc:481]: ok, I can extend this. But how is this for the other entity trait attributes? Because for those there is no need to use a `@GeneratedValue` annotation, right? How would you suggest to write it  then?

File: content/modules/rest/pages/entities-api.adoc (hunk 34 above)
Comment (mariodavid) [reply content/modules/rest/pages/entities-api.adoc]: see explanation above

**Comment by mariodavid** [comment]
@knstvk thanks for taking a look. Yes, passive voice is always a bad habit I oftentimes fall back into. I fixed the remarks except for the "NOTE" question. We can decide what we prefer for those out-of-band explanations like this parameter table.

File: content/modules/rest/pages/entities-api.adoc (hunk 22 above)
Comment (mariodavid) [reply content/modules/rest/pages/entities-api.adoc]: I escaped it for `((`. Now it seems to work.
Bildschirmfoto 2021-03-04 um 11 16 55

File: content/modules/rest/pages/entities-api.adoc (hunk 15 above)
Comment (knstvk) [reply content/modules/rest/pages/entities-api.adoc]: I think admonitions are not suitable here, even from the name. It's more a warning or an advice, right?
But here we have just a regular piece of information, a list of attributes or something.

How about using regular syntax for terms/definitions like that:
```
It is also possible to further control the behavior of the API by using the following URL query parameters:

fetchPlan:: name of an entity's xref:data-access:fetch-plans.adoc[fetch plan] `(String)`.
limit:: the number of entities to be returned by the API `(int)`.
offset:: the position of the first returned entity `(int)`.
sort:: an entity attribute that will be used for sorting. `+attribute` or `attribute` for ascending order, `-attribute` for descending order `(String)`.
```

It's rendered with decent visual differentiation:
image

File: content/modules/rest/pages/entities-api.adoc (hunk 31 above)
Comment (knstvk) [reply content/modules/rest/pages/entities-api.adoc:481]: Perhaps it depends on the trait.
For Audit and Soft Delete traits it's simple - the client should not send them.
The others are complicated:
- HasUUID trait includes id if it's of UUID type. Let's not mention it at all.
- Version should not be send on create but should be send on update otherwise optimictic lock won't work. Worth mentioning.

File: content/modules/rest/pages/entities-api.adoc [hunk 37]
```diff
@@ -0,0 +1,759 @@
...
+----
+
+@JmixEntity
+@Table(name = "RSTEX11_ORDER")
+@Entity(name = "rstex11_Order") //<1>
+public class Order {
+    // ...
+}
+----
+<1> the name attribute value `rstex11_Order` of the `@Entity` JPA annotation indicates the `entityName` parameter in the Entities API
```
Comment (knstvk) [inline content/modules/rest/pages/entities-api.adoc]: Also start from capital and dot at the end.

File: content/modules/rest/pages/entities-api.adoc (hunk 15 above)
Comment (mariodavid) [reply content/modules/rest/pages/entities-api.adoc]: you are right I think. I did the change accordingly 👍
